### PR TITLE
added elastic search client and backend

### PR DESCRIPTION
**Pull request by adityathebe**
Resolves https://github.com/flanksource/apm-hub/issues/18

## What's in here

- Updated from Go `1.17` to `1.20`
- Support backend routing
- Added support for opensearch as a search backend
- Fixed duration parser
- Lint fixes

## Edge cases (DRAFT)

> will create issues for these once this PR is ready

- Handling multiple next pages. The search result only contains one NextPage field however if multiple backends are used there can be multiple nextpages. (_Not handling this right now because most likely you're only going to use a single search backend_).
- A backend has multiple routes say (a, b and c) where **b** and **c** are not additive but **a** is. If a query matches all the routes, then only route a is considered _(because it's the first route we check against)_ and route b and c are basically discarded. In this case, even though the query matched route b and c, the results are not additive. So there's some anomalous behavior here. It's up to the config creator to carefully craft the config.

## Shortcomings
- Allowing nested fields to be part of the message fields. For example: Allow using `"log.file.path" `as the selector field for message.
- Allow exclusions on flattened labels? 
_Example_: If the search result has response `{"agent": {"name": "foo", "address": "bar"}} `It's not possible to exclude `agent.address` from the labels, you either have to exclude the entire `agent` top level field or include all.
- Exclusions and inclusions are supported but it's not possible to provide a list of them. _Example_: You need to exclude search results by agent name and you have 2 agents that you want to ignore. To do that, you'll have to add `"agent1": "nginx-ingress-controller-f6zx7"` and `"agent2": "nginx-ingress-controller-r46vg"` to your labels and access those fields in the query template. A better way would be to give a list and iterate through them in the query template.

## Example

For the given `samples/config-opensearch.yaml` config

```sh
curl -s -H "Content-type: application/json" \
-d '{"type": "elastic", "id": "elastic-backend", "limit": 1, "page": "", "labels": {"foo": "bar", "clientIP": "127.0.0.1", "requestURI": "/nginx-status"}, "start": "2w"}' \
localhost:8082/search | jq 
```

### CODE REVIEW

File: pkg/elasticsearch/search.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+		client:   client,
+		index:    index,
+		template: template,
+	}, nil
+}
+
+func (t *ElasticSearchBackend) Search(q *logs.SearchParams) (logs.SearchResults, error) {
+	var result logs.SearchResults
+	var buf bytes.Buffer
+	if err := t.template.Execute(&buf, q.Labels); err != nil {
```
Comment (adityathebe) [inline pkg/elasticsearch/search.go:39]: @moshloop I'm using Labels from the search param as the view for now.
I don't think any of the fields that we currently have in `logs.SearchParams` is suitable for this use case.
Should I add another field?

File: pkg/config.go [hunk 2]
```diff
@@ -52,7 +62,58 @@ func ParseConfig(kommonsClient *kommons.Client, configFile string) ([]logs.Searc
...
+			if err != nil {
+				return nil, fmt.Errorf("error getting the elastic search config: %w", err)
+			}
+
+			client, err := v8.NewClient(*cfg)
+			if err != nil {
+				return nil, fmt.Errorf("error creating the elastic search client: %w", err)
+			}
+
+			pingResp, err := client.Ping()
```
Comment (adityathebe) [inline pkg/config.go:77]: @moshloop Do we need to validate proper connection with the server before using it as a search backend?

Comment (moshloop) [reply pkg/config.go:77]: Yes, it makes sense

File: go.mod [hunk 3]
```diff
@@ -1,12 +1,12 @@
 module github.com/flanksource/flanksource-ui/apm-hub
 
 require (
+	github.com/elastic/go-elasticsearch/v8 v8.6.0
```
Comment (moshloop) [inline go.mod:4]: I think we will need add in support for the https://github.com/opensearch-project/opensearch-go/ client as well, as go-elasticsearch will not talk to OpenSearch

File: api/logs/logs.go [hunk 4]
```diff
@@ -40,6 +43,16 @@ type FileSearchBackend struct {
 	Paths  []string          `yaml:"path,omitempty"`
 }
 
+type ElasticSearchBackend struct {
+	Address  string `yaml:"address,omitempty"`
+	CloudID  string `yaml:"cloudID,omitempty"`
+	APIKey   string `yaml:"apiKey,omitempty"`
+	Username string `yaml:"username,omitempty"`
+	Password string `yaml:"password,omitempty"`
```
Comment (moshloop) [inline api/logs/logs.go:51]: These should all be kommons.EnvVarSource's

File: pkg/elasticsearch/search.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+		var (
+			id, _     = data["_id"].(string)
+			idx, _    = data["_index"].(string)
+			source, _ = data["_source"].(map[string]any)
+		)
+
+		resp = append(resp, logs.Result{
+			Id:      id,
+			Message: fmt.Sprintf("%v", source),
```
Comment (moshloop) [inline pkg/elasticsearch/search.go:91]: This will normally be the `message` or `log` or `msg` field,

Comment (moshloop) [inline pkg/elasticsearch/search.go:91]: Also need to map timestamp which is normally the `@timestamp` field

File: pkg/elasticsearch/search.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+func (t *ElasticSearchBackend) Search(q *logs.SearchParams) (logs.SearchResults, error) {
+	var result logs.SearchResults
+	var buf bytes.Buffer
+	if err := t.template.Execute(&buf, q.Labels); err != nil {
+		return result, fmt.Errorf("error executing template: %w", err)
+	}
+
+	res, err := t.client.Search(
```
Comment (moshloop) [inline pkg/elasticsearch/search.go:43]: I think we will need to handle paging

File: pkg/elasticsearch/search.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+			id, _     = data["_id"].(string)
+			idx, _    = data["_index"].(string)
+			source, _ = data["_source"].(map[string]any)
+		)
+
+		resp = append(resp, logs.Result{
+			Id:      id,
+			Message: fmt.Sprintf("%v", source),
+			Labels: map[string]string{
+				"index": idx,
```
[the remark this answers is not in the document]
Comment (adityathebe) [reply pkg/elasticsearch/search.go:93]: I only have access to _ fields.

image

File: pkg/elasticsearch/search.go (hunk 5 above)
Comment (adityathebe) [reply pkg/elasticsearch/search.go]: This will come from the config `yaml` file.

Comment (adityathebe) [reply pkg/elasticsearch/search.go]: This will come from the config yaml file.

File: pkg/elasticsearch/search.go (hunk 6 above)
Comment (adityathebe) [reply pkg/elasticsearch/search.go:44]: Added pagination

File: api/logs/logs.go [hunk 8]
```diff
@@ -6,26 +6,50 @@ import (
...
+	Path     string          `yaml:"-"`
 	Backends []SearchBackend `yaml:"backends,omitempty"`
 }
 
 type SearchBackend struct {
-	Routes     []SearchRoute            `json:"routes,omitempty"`
-	Backend    SearchAPI                `json:"-"`
-	Kubernetes *KubernetesSearchBackend `json:"kubernetes,omitempty"`
-	Files      []FileSearchBackend      `json:"file,omitempty" yaml:"file,omitempty"`
+	Routes        []SearchRoute            `json:"routes,omitempty"`
```
Comment (moshloop) [inline api/logs/logs.go:22]: I think this should go into a CommonBackend that all backends inline -  Have a route to multiple backends doesn't make much sense.

File: api/logs/logs.go [hunk 9]
```diff
@@ -6,26 +6,50 @@ import (
...
+	if t.Type != "" && t.Type != q.Type {
+		return false
+	}
+
+	if t.IdPrefix != "" && !strings.HasPrefix(q.Id, t.IdPrefix) {
+		return false
+	}
+
+	for k, v := range t.Labels {
+		if q.Labels[k] != v {
```
Comment (moshloop) [inline api/logs/logs.go:47]: I think should use the MatchItem format - i.e. split on "," - that would then enable us to support exsists `*`, not exists `!`

File: api/logs/logs.go (hunk 8 above)
Comment (adityathebe) [reply api/logs/logs.go:22]: `SearchBackend`, while it has fields for all the backends, it'll only ever be populated by one backend at max. So if I understand it's already like that.

Can you give me an example of how the config file should look like?

File: pkg/config.go [hunk 10]
```diff
@@ -1,34 +1,47 @@
...
 	if err := yaml.Unmarshal(data, searchConfig); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("error unmarshalling the configFile: %v", err)
 	}
 
+	return searchConfig, nil
+}
+
+// LoadBackendsFromConfig loads the backends from the config file
+func LoadBackendsFromConfig(kommonsClient *kommons.Client, searchConfig *logs.SearchConfig) ([]logs.SearchBackend, error) {
```
Comment (moshloop) [inline pkg/config.go:38]:
```suggestion
func LoadBackendsFromConfig(client *kommons.Client, searchConfig *logs.SearchConfig) ([]logs.SearchBackend, error) {
```

File: pkg/config.go [hunk 11]
```diff
@@ -52,7 +65,167 @@ func ParseConfig(kommonsClient *kommons.Client, configFile string) ([]logs.Searc
...
+			backend.Backend = es
+
+			backends = append(backends, backend)
+		}
 	}
 
 	return backends, nil
 }
+
+func getOpenSearchEnvVars(kClient *kommons.Client, conf *logs.OpenSearchBackend) (username, password string, err error) {
```
Comment (moshloop) [inline pkg/config.go:131]:
```suggestion
func getOpenSearchEnvVars(client *kommons.Client, conf *logs.OpenSearchBackend) (username, password string, err error) {
```

**Pull request merged**